core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -1854,6 +1857,11 @@ public void onSettingsChanged() {
             // the setting will be re-interpreted if it's set to true
             this.maxUnsafeAutoIdTimestamp.set(Long.MAX_VALUE);
         }
+        final TranslogDeletionPolicy translogDeletionPolicy = translog.getDeletionPolicy();
+        final IndexSettings indexSettings = engineConfig.getIndexSettings();
+        translogDeletionPolicy.setMaxRetentionAgeInMillis(indexSettings.getTranslogRetentionAge().getMillis());
+        translogDeletionPolicy.setRetentionSizeInBytes(indexSettings.getTranslogRetentionSize().getBytes());
+        translog.trimUnreferencedReaders();
```

Comment:
I don't think we should do this in the settings loop. if these settings changed and the user wants them to take effect they should call `_flush`